rust/src/collaboration/databasesync.rs

```diff
@@ -0,0 +1,708 @@
+        .ok_or(())
+}
+
+/// Download a file from its remote, saving all snapshots to a database in the
+/// specified location. Returns a FileContext for opening the file later.
+///
+/// * `file` - Remote File to download and open
+/// * `db_path` - File path for saved database
+/// * `progress` - Function to call for progress updates
+pub fn download_file<S: BnStrCompatible, F: ProgressCallback>(
+    file: &RemoteFile,
+    db_path: S,
+    mut progress: F,
+) -> Result<Ref<FileMetadata>, ()> {
+    let db_path = db_path.into_bytes_with_nul();
+    let result = unsafe {
+        BNCollaborationDownloadFile(
+            file.as_raw(),
+            db_path.as_ref().as_ptr() as *const ffi::c_char,
+            Some(F::cb_progress_callback),
+            &mut progress as *mut F as *mut ffi::c_void,
+        )
+    };
+    let success = !result.is_null();
+    success
+        .then(|| unsafe { Ref::new(FileMetadata::from_raw(result)) })
+        .ok_or(())
+}
+
+/// Upload a file, with database, to the remote under the given project
+///
+/// * `metadata` - Local file with database
+/// * `project` - Remote project under which to place the new file
+/// * `parent_folder` - Optional parent folder in which to place this file
+/// * `progress` - Function to call for progress updates
+/// * `name_changeset` - Function to call for naming a pushed changeset, if necessary
+pub fn upload_database<P: ProgressCallback, N: NameChangeset>(
+    metadata: &FileMetadata,
+    project: &RemoteProject,
+    parent_folder: Option<&RemoteFolder>,
+    mut progress: P,
+    mut name_changeset: N,
+) -> Result<RemoteFile, ()> {
+    let folder_raw = parent_folder.map_or(ptr::null_mut(), |h| unsafe { h.as_raw() } as *mut _);
+    let result = unsafe {
+        BNCollaborationUploadDatabase(
+            metadata.handle,
+            project.as_raw(),
+            folder_raw,
+            Some(P::cb_progress_callback),
+            &mut progress as *mut P as *mut ffi::c_void,
+            Some(N::cb_name_changeset),
+            &mut name_changeset as *mut N as *mut ffi::c_void,
+        )
+    };
+    ptr::NonNull::new(result)
+        .map(|raw| unsafe { RemoteFile::from_raw(raw) })
+        .ok_or(())
+}
+
+/// Test if a database is valid for use in collaboration
+pub fn is_collaboration_database(database: &Database) -> bool {
+    unsafe { BNCollaborationIsCollaborationDatabase(database.as_raw()) }
+}
+
+/// Get the Remote for a Database
+pub fn get_remote_for_local_database(database: &Database) -> Result<Option<Remote>, ()> {
+    let mut value = ptr::null_mut();
+    let success =
+        unsafe { BNCollaborationGetRemoteForLocalDatabase(database.as_raw(), &mut value) };
+    success
+        .then(|| ptr::NonNull::new(value).map(|handle| unsafe { Remote::from_raw(handle) }))
+        .ok_or(())
+}
+
+/// Get the Remote for a BinaryView
+pub fn get_remote_for_binary_view(bv: &BinaryView) -> Result<Option<Remote>, ()> {
```

Comment:
`Result<Option<Remote>, ()>` -> `Option<Remote>`

Reply:
This need to be `Result<Option<T>, ()>` or `Option<Option<T>>`. The user need to differentiate an error (possible from the network) from the local database being local only and simply having `None` remote.